# MasterThesis

Analysis of single-cell RNA-seq data for the Single Cell Analysis (SIGA) group of the Helmholtz Centre for Infection Research ([HZI](https://www.helmholtz-hzi.de/en/)) Institute for RNA-based Infection Research ([HIRI](https://www.helmholtz-hiri.de/)). 

This public repository contains the source code that has been used to analyze the data. 

## Projects:
1. Transcriptional Profiling of the Immune Response to Influenza Vaccination
2. Dissecting the Cellular Composition of Pediatric Kidney Tumors
3. Exploring the Transcriptional Regulation in *Trypanosoma brucei*
4. Dissecting the Microenvironment of the Squamo-Columnar Junction of the Cervix
5. Exploring the Cellular Repertoire of Myeloid-Derived Suppressor Cells

## Workflows
The data analysis frameworks, workflows and tools that were used to develop this script are [Seurat](https://satijalab.org/seurat/), [Scanpy](https://scanpy.readthedocs.io/en/stable/), [Scran](https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html) and the [single-cell RNA-seq course](https://hemberg-lab.github.io/scRNA.seq.course/index.html) from the University of Cambridge. 

### Chromium droplet-sequencing data analysis
The script for the analysis of Drop-Seq data is currently based on the [Seurat](https://satijalab.org/seurat/) workflow and executed in [R](https://www.r-project.org/). 
